lib/index.js

```diff
@@ -0,0 +1,240 @@
+api.consensusMethod = 'Continuity2017';
+
+api.type = 'VeresOne';
+
+// register this ledger plugin
+bedrock.events.on('bedrock.start', () => brLedgerNode.use(
+  'VeresOne', {api, type: 'electorSelection'}));
+
+api.getBlockElectors = async ({
+  ledgerNode, ledgerConfiguration, latestBlockSummary, blockHeight,
+  recoveryMode = false
+}) => {
+  // TODO: a cache may be added here to prevent running
+  // `_getElectorPoolElectors` when recoveryMode switches from false to true.
+  // There is already a cache in Continuity that prevents the `getBlockElectors`
+  // API from being called more than once with the same `blockHeight` and
+  // `recoverMode` values. Since recoveryMode should be rare, a cache may
+  // be unncessary
+  const electorPoolElectors = Object.values(
+    await brLedgerUtils.getElectorPoolElectors(
+      {ledgerConfiguration, ledgerNode}));
+  let electors = await api._computeElectors({
+    blockHeight, electorPoolElectors, latestBlockSummary, ledgerConfiguration,
+    ledgerNode
+  });
+
+  let recoveryGenerationThreshold;
+
+  let recoveryElectors = api._computeRecoveryElectors(
+    {blockHeight, electors, electorPoolElectors});
+
+  if(recoveryMode) {
+    electors = api._computeElectorsForRecoveryMode(
+      {blockHeight, electors, recoveryElectors});
+    recoveryElectors = [];
+  } else if(electors.length > 1) {
+
+    // FIXME: some or all of this is no longer required?
+
+    // const minimumMergeEvents = api._computeRecoveryMinimumMergeEvents({f});
+    const minimumMergeEvents = 20;
+    // allow up to 10x inefficiency before trying to recover
+    recoveryGenerationThreshold = minimumMergeEvents * 10;
+  }
+
+  logger.verbose(
+    'Selected Electors:',
+    {ledgerNode: ledgerNode.id, blockHeight, electors, recoveryElectors});
+
+  return {
+    // number of electors, must be 3f+1 (or 1 for dictator model), safe
+    // for up to `f` byzantine failures
+    electors,
+    // optional `f+1` "trusted" electors that must not fail and can handle
+    // catastrophic case where remaining `2f` electors cannot communicate
+    // (recovery electors are not a failsafe against a coordinated byzantine
+    // attack as the byzantine electors could collude to show different sides
+    // of a fork, but safe against network failures preventing consensus)
+    recoveryElectors,
+    // number of generations of merge events where no new electors participate
+    // and fewer than 2f+1 total electors (including self) participate causing
+    // recovery mode to trigger because consensus cannot be reached
+    recoveryGenerationThreshold,
+  };
+};
+
+// it is useful to override this function in tests
+api._computeElectors = async ({
+  blockHeight, electorPoolElectors, latestBlockSummary, ledgerConfiguration,
+  ledgerNode
+}) => {
+  // get partipicants for the last block
+
+  // FIXME: if we're using the _blocks API here, should it be made public
+  // (remove the underscore?)
```

Comment:
Sounds like it should be made public, but I don't know what it would look like. What's the proposal?

Reply:
The proposal is to remove the underscore from `_block`.  Since you don't seem to be concerned about it, I say we skip it.

Reply:
We can skip it for now. We should come up with a clean public API for it later.